v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -313,6 +386,76 @@ def set_credentials(self, username, password=None):
         """
         self._mqtt_client.username_pw_set(username=username, password=password)
 
+    def add_incoming_message_filter(self, topic):
+        """
+        Filter incoming messages on a specific topic.
+
+        :param str topic: The topic you wish to filter on
+
+        :raises: ValueError if a filter is already applied for the topic
+        """
+        if topic in self._incoming_filtered_messages:
+            raise ValueError("Filter already applied for this topic")
+
+        # Add a Queue for this filter
+        self._incoming_filtered_messages[topic] = asyncio.Queue()
+
+        # Define a callback that uses the filter queue
+        self_weakref = weakref.ref(self)
```

Comment:
I don't think this weakref stuff is still necessary, is it?

Reply:
Last time I checked early in V3 it was still necessary under some circumstances. That said, it's worth revisiting once I have a full E2E suite set up.